lib/core/test/unit/Cardano/Pool/RankingSpec.hs

```diff
@@ -0,0 +1,254 @@
+    = PositiveEffect
+    | NegativeEffect
+    deriving (Show, Eq)
+
+-- | All else equal, what effect on the desirability does changing a single
+-- field have?
+allElseEqualProperty
+    :: forall (field :: Symbol) a.
+        ( Arbitrary a
+        , Show a
+        , Ord a
+        , HasField' field Pool a
+        , KnownSymbol field
+        )
+    => EffectOnDesirability
+        -- ^ Expected effect on the desirability
+    -> Pool
+       -- ^ The base pool which is later modified
+    -> EpochConstants
+       -- ^ Constants used
+    -> Property
+allElseEqualProperty expectedEffect pool constants =
+    forAll arbitrary $ \(value1 :: a) (value2 :: a) -> do
+        let (poolLower, poolHigher) = sortTuple (value1, value2)
+                & bimap poolWith poolWith
+
+        let (dLower, dHigher) =
+                ( desirability constants poolLower
+                , desirability constants poolHigher
+                )
+
+        let poolRankDesc = unlines
+                [ "Pools:"
+                , showPool poolLower  dLower
+                , showPool poolHigher dHigher
+                ]
+
+        case dLower `compare` dHigher of
+            EQ -> property True
+            -- A reason we end up in this case could be that both pools are
+            -- un-profitable. It would be interesting to re-run tests with
+            -- a higher total reward pot. But as long we hit the other cases we
+            -- should be fine, really.
+            GT -> expectedEffect === PositiveEffect
+            LT -> expectedEffect === NegativeEffect
+          & counterexample poolRankDesc
+          & classify (dLower == dHigher) "same desirability"
+  where
+    poolWith :: a -> Pool
+    poolWith val = pool & (field' @field) .~ val
+
+    sortTuple :: (a, a) -> (a, a)
+    sortTuple (a, b)
+        | a > b     = (a, b)
+        | otherwise = (b, a)
+
+    showPool :: Pool -> Double -> String
+    showPool x d = unwords
+        [ show (symbolVal $ Proxy @field)
+        , "="
+        , show (x ^. field' @field)
+        , ", desirability ="
+        , show d
+        ]
+
+instance Arbitrary Pool where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+instance Arbitrary (RelativeStakeOf a) where
+    arbitrary =
+        RelativeStake <$> choose (0, 1)
+
+    shrink (RelativeStake x) =
+        map RelativeStake $ filter (\a -> a >= 0 && a <= 1) $ shrink x
+
+-- TODO: We should ideally not export the NonNegative and Positive constructors,
+-- in which case we wouldn't be able to use derivingVia.
+deriving via (NonNegative Double) instance (Arbitrary Lovelace)
+deriving via (NonNegative Double) instance (Arbitrary (R.NonNegative Double))
+deriving via (Positive Int) instance (Arbitrary (R.Positive Int))
+
+instance Arbitrary Margin where
+    arbitrary = unsafeMkMargin <$> choose (0, 1)
+    shrink = map unsafeMkMargin
+        . filter (\a -> a >= 0 && a <= 1)
+        . map getMargin
+        . shrink
```

Comment:
The arbitrary instances for `Margin` and `RelativeStakeOf` are identical which re-enforce the argument that these should really be of the same type.